types/module.go

```diff
@@ -0,0 +1,219 @@
+package types
+
+import (
+	"encoding/json"
+
+	"github.com/spf13/cobra"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// ModuleClients helps modules provide a standard interface for exporting client functionality
+type ModuleClients interface {
+	GetQueryCmd() *cobra.Command
+	GetTxCmd() *cobra.Command
+}
+
+// AppModule is the standard form for an application module
+type AppModule interface {
+
+	// app name
+	Name() string
+
+	// registers
+	RegisterInvariants(InvariantRouter)
+
+	// routes
+	Route() string
+	NewHandler() Handler
+	QuerierRoute() string
+	NewQuerierHandler() Querier
+
+	//// genesis
+	//DefaultGenesisState() json.RawMessage
+	//ValidateGenesis(json.RawMessage) error
+	InitGenesis(Context, json.RawMessage) ([]abci.ValidatorUpdate, error)
+	//ExportGenesis(Context) json.RawMessage
+
+	BeginBlock(Context, abci.RequestBeginBlock) Tags
+	EndBlock(Context, abci.RequestEndBlock) ([]abci.ValidatorUpdate, Tags)
+}
+
+// module manananaager
+type ModuleManager struct {
+	Modules            map[string]AppModule
+	OrderInitGenesis   []string
+	OrderExportGenesis []string
+	OrderBeginBlockers []string
+	OrderEndBlockers   []string
+}
+
+// NewModuleManager creates a new ModuleManager object
+func NewModuleManager(modules ...AppModule) *ModuleManager {
+
```

Comment:
Maybe the module manager wants to store the `[]string` of the order in which the modules were passed in and save it as a default for some of the `SetOrder*` functions?

Reply:
Hmm good idea! - this will save on if statement's latter on